### ISSUE TITLE
Very well done!

### ISSUE THREAD

**Issue by Zibri** [subversivo58/lab#2]
How did you create "webpush-browserify-bundle.js" ?
Can you post a 2 lines guid to understand it?
I tried to use browserify on "web-push" package but gave me a lot of errors.

**Comment by subversivo58** [comment]
I did this to test and help you in question [351](https://github.com/web-push-libs/web-push/issues/351) because tbm was my doubt since I had tried before without success.

You should take into account that the script is out of date compared to the current version of the main library (webpush).

Step by step is simple (assuming you already have Browserify installed):

**1-** create an **.js** file ... in this example I'll use "main.js" (You can name it any way you want)

**2-** add the following code:

```javascript
window.WebPushLib = require('web-push')
```
When Browserify wrap it will add the **webpush** library object in the window object

**3-** open your terminal at the location of your file (main.js) and type the commands for Broserify:
```code
browserify  main.js -o webpush-browserify-bundle.js
```

**4-** include the script generated in your document to be able to use it, example:

```html
<!DOCTYPE html>
<head>
    <meta charset="utf-8">
</head>
<body>
    <script src="path/to/webpush-browserify-bundle.js"></script>
    <script>
        // teste if "bundled" webpush lib have in `window` object
        console.log(window.WebPushLib) // or 'WebPushLib' in window
     </script>
</body>
</html>
```

From this point we can use the library normally.

I hope I have helped you, enjoy

-----------------------

PS: [Broserify reference link](http://browserify.org/)

**Comment by subversivo58** [comment]
If you want to use a proxy to bypass the FCM lock find the prototype:

```javascript
// in "webpush-browserify-bundle.js" file
ClientRequest.prototype._onFinish = function () {}
```

Within this prototype change the destination url by sending to your proxy, do this before the `global.fetc ()` statement.

It should look like this:

```javascript
// test if URI is FCM
if ( /https:\/\/fcm\.googleapis\.com/.test(self._opts.url) ) {
    // change for your proxy
    self._opts.url = 'https://url-for-your-proxy-route/' + self._opts.url
}

// before this instruction
global.fetch(self._opts.url, {...})
```

In my original example I used the demonstration proxy library "cors-anywhere":

```javascript
if ( /https:\/\/fcm\.googleapis\.com/.test(self._opts.url) ) {
    self._opts.url = 'https://cors-anywhere.herokuapp.com/' + self._opts.url
}
```
Be aware when using this demo because the author himself recommends using his own proxy because there is no guarantee of maintenance of this demo nor its constant availability (Heroku apps "sleep" after some time of inactivity)

**Comment by Zibri** [comment]
yep! I knew all this... butmy mistake was that I was using some broken browserify lite.....

I also tried this:

`browserify -r web-push -s WebPushLib -o webpush-browserify-bundle.js`

no need for the fake main.js

**Comment by Zibri** [comment]
Also I wonder (and I am almost sure) if there is a way not to use a proxy...
I use my own on google servers but for a project I am doing I'd rather find a way not to use proxies at all.
I know I can use firebase and my project works well with that but I wish to make totally independent (i.e. unblockable)

**Comment by Zibri** [comment]
By the way, check this page https://gauntface.github.io/simple-push-demo/
In developer mode/network you can see the XHR request it does... and it works cross origin :smile: 
But still it would be better to find a way not to rely on additional servers otherwise this is all useless and it's better to use firebase as I already did (to send/receive push messages from a static html page or web-app)

**Comment by Zibri** [comment]
```javascript
fetch("https://simple-push-demo.appspot.com/api/v2/sendpush", {
    "credentials":"omit",
    "headers":{
        "content-type":"text/plain;charset=UTF-8"
    },
    "referrer":"https://gauntface.github.io/simple-push-demo/",
    "referrerPolicy":"no-referrer-when-downgrade",
    "body":"{\"headers\":{\"TTL\":10,\"Content-Length\":0,\"Authorization\":\"vapid t=eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJhdWQiOiJodHRwczovL2ZjbS5nb29nbGVhcGlzLmNvbSIsImV4cCI6MTU2MjY5OTg0Nywic3ViIjoibWFpbHRvOnNpbXBsZS1wdXNoLWRlbW9AZ2F1bnRmYWNlLmNvLnVrIn0.4pwePMiJ2DJFzwvzBNJO4rOotPcArGS6O1-psgSjxvluPf4OJxxbcQhVAqATgy9gfDESSMYZYoyR9k50Xk4uEg, k=BDd3_hVL9fZi9Ybo2UUzA284WG5FZR30_95YeZJsiApwXKpNcF1rRPF3foIiBHXRdJI2Qhumhf6_LFTeZaNndIo\"},\"endpoint\":\"https://fcm.googleapis.com/wp/dHDjDjdOOHk:APA91bHsqaIf2c1X8267DPuHO0RIJDTv4AOwfWzhcmzp1tJZsJ6aZOY4FuP_MbUZncxNqoPG__gEnb51uJT0iXl6EPIK3OpECX_2pVc_iu7u-PmtAm4gDNKsq-4HO2NKTntLdTrvCJat\"}",
    "method":"POST",
    "mode":"cors"
});
```

**Comment by Zibri** [comment]
Anyway, again.. using some clever code on google cloud (the free version) or any other proxy, works but if I have to use that then I prefer to use firebase directly... though I am almost sure there must be a way to POST to fcm.googleapis.com without needing a proxy.

**Comment by subversivo58** [comment]
> yep! I knew all this... butmy mistake was that I was using some broken browserify lite.....
> 
> I also tried this:
> 
> `browserify -r web-push -s WebPushLib -o webpush-browserify-bundle.js`
> 
> no need for the fake main.js

I confess that I have never used Browserify so I believe that it is to be used in the format **comom.js** or **require** ... I (preferentially) like to add to the global object `window` so I use a reference file (main.js).

**Comment by subversivo58** [comment]
> Anyway, again.. using some clever code on google cloud (the free version) or any other proxy, works but if I have to use that then I prefer to use firebase directly... though I am almost sure there must be a way to POST to fcm.googleapis.com without needing a proxy.

If you use in a desktop application (NW.js or Electron) for example, you could use `cURL`.

Now if you want to use "web-only" between browsers, the problem is that all browsers send **cors** headers in cross-origin requests, the server **FCM** does not accept the request when it sees a **cors header** ... take a look at this answer on [stackoverflow.com](https://stackoverflow.com/questions/36691533/how-to-send-push-notifications-in-chromeprogressive-web-apps/36717282#36717282)

This is why you need to use a proxy if you are trying to do this on the web (from the browser to the FCM endpoint) to bypass this block.

I find it easier to create a proxy app on [heroku.com](https://www.heroku.com) or [glitch.com](https://glitch.com) than to use firebase ... and at the end of those accounts a proxy app only forwards the request, it does not have to "handle" anything.

**Comment by subversivo58** [comment]
> Now if you want to use "web-only" between browsers, the problem is that all browsers send cors headers in cross-origin requests, the server FCM does not accept the request when it sees a cors header ... take a look at this answer on stackoverflow.com

I wanted to clarify this point ... because I myself re-read and found it unclear.

Because the web-push library defines custom headers other than "common" headers (4 headers in total)

```javascript
(4) [Array(2), Array(2), Array(2), Array(2)]
0: (2) ["TTL", 60]
1: (2) ["Content-Type", "application/octet-stream"]
2: (2) ["Content-Encoding", "aes128gcm"]
3: (2) ["Authorization", "vapid t=xxxxxxx.yyyyyyyy"] // omited in this example
length: 4
__proto__: Array(0)
```

The browser (all browsers as far as I know) first send an **OPTIONS** method request.

And it is this request (OPTIONS) that fails on FCM.

I imagine FCM only validates a request with **OPTIONS** method if it has the headers used in a "firebase app" ... you can [see this documentation](https://firebase.google.com/docs/cloud-messaging/http-server-ref).

**Comment by Zibri** [comment]
> > I also tried this:
> > `browserify -r web-push -s WebPushLib -o webpush-browserify-bundle.js`
> > no need for the fake main.js
> 
> I confess that I have never used Browserify so I believe that it is to be used in the format **comom.js** or **require** ... I (preferentially) like to add to the global object `window` so I use a reference file (main.js).

That's exactly what "-s WebPushLib" does.

**Comment by Zibri** [comment]
@subversivo58 please drop me an email (zibri AT zibri DOT org) so I can explain better (here is not the right place for that)

**Comment by Zibri** [comment]
anyway the purpose is to find a way to make 2 browser communicate without needing a server.. webrtc is a way... but since the "notification server" can't be easily taken down, also that is a very good option.
I can tell you a NUMBER of other ways using a second server... there is even no need to set up your own anywhere... My purpose was to use notifications and nothing else.
Sure that can be done by disabling XSS on browsers or by local applications. The difficult and so "funny" way would be to be able to do from within a normal browser and nothing else that could be "taken down".

**Comment by Zibri** [comment]
Shorter way:

```
browserify -r web-push -s WebPushLib -o WebPushLib.js
sed -i "s,self._opts.url,'https://cors-anywhere.herokuapp.com/'+self._opts.url," WebPushLib.js
```
